src/JustSaying/Messaging/Middleware/Backoff/BackoffMiddlewareBuilderExtensions.cs

```diff
@@ -0,0 +1,24 @@
+using System;
+using JustSaying.Messaging.MessageProcessingStrategies;
+using JustSaying.Messaging.Middleware.ErrorHandling;
+using JustSaying.Messaging.Monitoring;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Middleware.Backoff
+{
+    public static class BackoffMiddlewareBuilderExtensions
+    {
+        public static HandlerMiddlewareBuilder UseBackoff(this HandlerMiddlewareBuilder builder)
+        {
+            if (builder == null) throw new ArgumentNullException(nameof(builder));
+
+            var backoffStrategy = builder.ServiceResolver.ResolveOptionalService<IMessageBackoffStrategy>();
+            if (backoffStrategy == null) return builder;
```

Comment:
Yeah it's a fair question. We could instead do the check in `UseDefaults` to determine whether to call this, and then throw if we can't find an instance of `IMessageBackoffStrategy`? I tended towards the way it is but don't have strong feelings.

```csharp
builder.UseMessageContextAccessor();
if (builder.ServiceResolver.ResolveOptionalService<IMessageBackoffStrategy>() != null)
{
    builder.UseBackoff();
}
builder.UseErrorHandler();
builder.Use<LoggingMiddleware>();
builder.UseStopwatch(handlerType);
builder.Use<SqsPostProcessorMiddleware>();
builder.UseHandler<TMessage>();
```

Reply:
I don't have strong feelings on a solution, it just seems that if something was misconfigured, if it didn't throw or something, then it would be less surprising/fail-fast than if it just silently looked like it had done something.

Reply:
That's a fair point I think - the failure behaviour shouldn't really be nothing. I thought about a log but it would log by default if no back off was registered, which is the default.

I also experimented with a `UseIf` function that conditionally added it to the pipe, but it stuffs with the ordering and makes things a bit too unpredictable I think, so pulled it out.

Perhaps another option might be to require the user to add the backoff themselves? Instead of auto adding it if we find an IMessageBackoffStrategy, we could flip it so the user has to do `builder.UseBackoff(IMessageBackoffStrategy strategy)`. Feels a bit service locatorey to me though as you'd need to pull it out or create it separately beforehand.

Reply:
So I've been going around a bit on this one, given its the last bit of the puzzle 😅. I'm leaning a bit towards removing `IMessageBackoffStrategy` from DI and instead requiring the user to pass it in to the middleware, like so:

```csharp
builder.UseBackoff(IMessageBackoffStrategy strategy)
```

And I think this has a few benefits:
- It would be possible to use a different backoff strategy for different subscriptions, whereas at the moment there's only one in DI.
- It'd mean we can leave the call to `UseBackoff` out of the defaults, and users would opt-in as above.
- We can check for a null IMBS in UseBackoff

thoughts?

Reply:
Making this opt-in seems fair. We could always have it as a default in JustSayingStack.

Reply:
Yeah I like this suggestion of opting in at the middleware builder layer

Reply:
I've PR'd this approach here to make the change clear: https://github.com/gkinsman/JustSaying/pull/153.

There is a downside, which is that we'd now have to always receive the `ApproximateReceiveCount` attribute as before we only fetched it if an IMBS was registered.

Thoughts @slang25 @martincostello?

Reply:
Just thinking out loud: could there be an internal config-y/options-y type that could be queried to control whether or not to fetch it, then adding back-off would add that in and make it get configured to be fetched at the appropriate place?

Reply:
Ah that's an interesting idea - perhaps a set of receive attributes that is configurable. That would open the doors to future middlewares that wanted to use attributes other than just approx receive count.

I think if we configure it on the `ISqsQueue` then it could be used by the receive buffer, and set by the backoff middleware - there might be some trickery but will see if it falls out.

Reply:
How would we feel about merging this as-is and following up with the optimisation to prevent always loading the approx count? Cautious that this is already new-concept heavy 🙂

Reply:
Works for me.